dali/sasha/saxref.cpp

```diff
@@ -670,6 +678,61 @@ class CNewXRefManagerBase
 
     }
 
+    void startHeartbeat(const char * op)
+    {
+        CriticalBlock block(heartbeatSect);
+        time(&heartbeatStartTime);
+        lastHeartbeat = heartbeatStartTime;
+        heartbeatInterval = 60; // Start with 1 minute (in seconds)
+        heartbeatCount = 0;
+        uncondlog("%s heartbeat started (interval: 1 minute)", op);
+    }
+
+    void checkHeartbeat(const char * op)
+    {
+        CriticalBlock block(heartbeatSect);
+        time_t now = time(NULL);
+
+        if ((now - lastHeartbeat) >= heartbeatInterval) {
+            unsigned elapsedMinutes = (unsigned)(now - heartbeatStartTime) / 60;
+            unsigned elapsedHours = elapsedMinutes / 60;
+            unsigned remainingMinutes = elapsedMinutes % 60;
+
+            heartbeatCount++;
+
+            // Format UTC timestamp directly from time_t
+            struct tm *utc_tm = gmtime(&now);
+            char timestamp[32];
+            strftime(timestamp, sizeof(timestamp), "%Y-%m-%d %H:%M:%S", utc_tm);
+
+            if (elapsedHours > 0) {
+                log("%s - elapsed: %uh %um (%s UTC)", op, elapsedHours, remainingMinutes, timestamp);
+            } else {
+                log("%s - elapsed: %um (%s UTC)", op, elapsedMinutes, timestamp);
+            }
+
+            if (heartbeatCount > 1) {
+                unsigned logCount = 0;
```

Comment:
I suspect this logic could be simplified.  It should at least have a comment to describe the logic.